presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java

```diff
@@ -399,6 +425,10 @@ private void freePartitions()
         @Override
         public void close()
         {
+            LookupSource lookupSource = suppliedLookupSources.remove(this);
```

Comment:
Looks like this is the only unguarded access, maybe just lock here and make suppliedLookupSources non concurrent.